RecoEgamma/EgammaTools/plugins/HGCalElectronIDValueMapProducer.cc

```diff
@@ -0,0 +1,238 @@
+// -*- C++ -*-
+//
+// Package:   RecoEgamma/HGCalElectronIDValueMapProducer
+// Class:    HGCalElectronIDValueMapProducer
+//
+/**\class HGCalElectronIDValueMapProducer HGCalElectronIDValueMapProducer.cc RecoEgamma/HGCalElectronIDValueMapProducer/plugins/HGCalElectronIDValueMapProducer.cc
+
+ Description: [one line class summary]
+
+ Implementation:
+    [Notes on implementation]
+*/
+//
+// Original Author:  Nicholas Charles Smith
+//      Created:  Wed, 05 Apr 2017 12:17:43 GMT
+//
+//
+
+
+// system include files
+#include <memory>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/Common/interface/ValueMap.h"
+
+#include "DataFormats/EgammaCandidates/interface/GsfElectron.h"
+#include "DataFormats/EgammaCandidates/interface/GsfElectronFwd.h"
+#include "DataFormats/GsfTrackReco/interface/GsfTrack.h"
+
+#include "RecoEgamma/EgammaTools/interface/ElectronIDHelper.h"
+#include "RecoEgamma/EgammaTools/interface/LongDeps.h"
+
+class HGCalElectronIDValueMapProducer : public edm::stream::EDProducer<> {
+  public:
+    explicit HGCalElectronIDValueMapProducer(const edm::ParameterSet&);
+    ~HGCalElectronIDValueMapProducer() override;
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  private:
+    void beginStream(edm::StreamID) override;
+    void produce(edm::Event&, const edm::EventSetup&) override;
+    void endStream() override;
+
+    // ----------member data ---------------------------
+    edm::EDGetTokenT<edm::View<reco::GsfElectron>> ElectronsToken_;
+    float radius_;
+    std::map<const std::string, std::vector<float>> maps_;
+
+    std::unique_ptr<ElectronIDHelper> eIDHelper_;
+};
+
+HGCalElectronIDValueMapProducer::HGCalElectronIDValueMapProducer(const edm::ParameterSet& iConfig) :
+  ElectronsToken_(consumes<edm::View<reco::GsfElectron>>(iConfig.getParameter<edm::InputTag>("electrons"))),
+  radius_(iConfig.getParameter<double>("pcaRadius"))
+{
+  // All the ValueMap names to output are defined in the python config
+  // so that potential consumers can configure themselves in a simple manner
+  for(auto key : iConfig.getParameter<std::vector<std::string>>("variables")) {
+    maps_[key] = {};
+    produces<edm::ValueMap<float>>(key);
```

Comment:
why is ```variables``` configurable?
all values are hardcoded in the code and there is even an exception if there is a mismatch.
The keys should be then hardcoded as well, and perhaps even mapped to enums to avoid the costly string manipulations.

Reply:
Is there a good reason to put about 40 values into individual float value maps?
It seems more productive to make a persistable struct for these

Reply:
The point was to make the configuration easy for a rapidly changing set of variables, and in particular for the eventual consumer module at MiniAOD level that will take these floats and return a MVA score.
Eventually some subset of these variables should end up in a struct in the `reco::` objects themselves as they are a core feature of the HGCal objects.  For now we have tried to go the route of least resistance due to the time pressure of the TDR.

Reply:
... looking back at the configs, it looks like the idea is to stuff these back into user floats at the PAT level.
In this case perhaps the individual float value maps are simpler to handle.

Reply:
I still stay with my original comment about hardcoding the keys.

If they need to be exposed to pat userFloats, the list of keys can be filled into python via fillDescriptions

Reply:
this comment is still not addressed